arangod/RestHandler/RestAdminClusterHandler.cpp

```diff
@@ -881,6 +885,142 @@ RestStatus RestAdminClusterHandler::handleQueryJobStatus() {
+  } else {
+    generateError(rest::ResponseCode::BAD, TRI_ERROR_BAD_PARAMETER, "object with key `id`");
+    return RestStatus::DONE;
+  }
+
+  auto targetPath = arangodb::cluster::paths::root()->arango()->target();
+  std::vector<std::string> paths = {
+      targetPath->pending()->job(jobId)->str(),
+      targetPath->failed()->job(jobId)->str(),
+      targetPath->finished()->job(jobId)->str(),
+      targetPath->toDo()->job(jobId)->str(),
+  };
+
+  try {
+    auto& agencyCache = server().getFeature<ClusterFeature>().agencyCache();
+    auto [acb, idx] = agencyCache.read(paths);
+    auto res = acb->slice();
+
+    auto paths = std::vector{
+      targetPath->pending()->job(jobId)->vec(),
+      targetPath->failed()->job(jobId)->vec(),
+      targetPath->finished()->job(jobId)->vec(),
+      targetPath->toDo()->job(jobId)->vec(),
+    };
+
+    for (auto const& path : paths) {
+      VPackSlice job = res.at(0).get(path);
+
+      if (job.isObject()) {
+        LOG_DEVEL << job.toJson();
+        LOG_TOPIC("eb139", INFO, Logger::SUPERVISION)
+          << "Attempting to abort supervision job " << job.toJson();
+        auto type = job.get("type").copyString();
+        if (type != "moveShard") { // only moveshard may be aborted
+          generateError(Result{400, "Only MoveShard jobs can be aborted"});
+          return RestStatus::DONE;
+        } else if (path[2] != "Pending" && path[2] != "ToDo") {
+          generateError(Result{400, path[2] + " jobs can no longer be cancelled."});
+          return RestStatus::DONE;
+        }
+
+        auto sendTransaction = [&] {
+          VPackBuffer<uint8_t> trxBody;
+          { VPackBuilder builder(trxBody);
+            { VPackArrayBuilder trxs(&builder);
+              if (path[2] == "ToDo") {
+                VPackArrayBuilder trx(&builder);
+                { VPackObjectBuilder op(&builder);
+                  builder.add("arango/Target/ToDo/" + jobId + "/abort", VPackValue(true)); }
+                { VPackObjectBuilder pre(&builder);
+                  builder.add(VPackValue("arango/Target/ToDo/" + jobId));
+                  { VPackObjectBuilder val(&builder);
+                    builder.add("oldEmpty", VPackValue(false)); }}
+              }
+              VPackArrayBuilder trx(&builder);
+              { VPackObjectBuilder op(&builder);
+                builder.add("arango/Target/Pending/" + jobId + "/abort", VPackValue(true)); }
+              { VPackObjectBuilder pre(&builder);
+                builder.add(VPackValue("arango/Target/Pending/" + jobId));
+                { VPackObjectBuilder val(&builder);
+                  builder.add("oldEmpty", VPackValue(false)); }}
+            }
+          }
+          return AsyncAgencyComm().sendWriteTransaction(60s, std::move(trxBody));
+        };
+
+        waitForFuture(
+          sendTransaction()
+          .thenValue([this](AsyncAgencyCommResult&& wr) {
+            if (!wr.ok()) {
+              if (wr.statusCode() == 412) {
```

Comment:
If the job is in ToDo according to the cache, and then it is moved to pending, then we run into a precondition failed and report an error. This is not really correct. Can we maybe do a retry in this case? This time in pending? If we arranged the loop to first look into ToDo and only then in Pending, this might be doable?

Reply:
i abort both jobs. pending and todo